### ISSUE TITLE
"Fill struct fields" code action inside proc_macro invocation is broken by iterating the TokenStream

### ISSUE THREAD

**Issue by LHolten** [rust-lang/rust-analyzer#16425]
I hope this is the right place for this report, I have asked on the [forum](https://users.rust-lang.org/t/tokenstream-iteration-is-breaking-rust-analyzer-code-action-is-this-a-bug-in-the-compiler/105746) first.

## How to reproduce
First I define two proc macros that should both just return the input stream.

```
#[proc_macro]
pub fn identity_correct(input: TokenStream) -> TokenStream {
    input
}

#[proc_macro]
pub fn identity_broken(input: TokenStream) -> TokenStream {
    let mut out = TokenStream::new();
    out.extend(input.into_iter());
    out
}
```

Then I use these proc macros on a struct defintion that is incomplete:

```
struct Test {
    x: u32,
    y: u32,
};

fn main() {
    identity_correct!(Test { x: 0 });
    identity_broken!(Test { x: 0 });
}
```

When I then apply the rust-analyzer "Fill struct fields" code action on both I get the following result:

```
fn main() {
    identity_correct!(Test {x:0, y: todo!() });
    identity_broken!(Test {x:0, y: todo!() } });
}
```

Note that there is one extra closing brace `}` for the broken macro.

## Conclusion
Clearly `out.extend(input.into_iter())` loses some information and rust-analyzer can not handle it.
Since I don't use any external dependencies, this could be a bug in rust-analyzer or rust.

**rust-analyzer version**: 0.3.1815-standalone

**rustc version**: 1.77.0-nightly

**Comment by Veykril** [comment]
Hm, looks like we might mixing up syntax contexts somewhere here.

**Comment by ihor-rud** [comment]
I am experiencing the same issue. For example this code 
```rust
#[tracing::instrument]
pub fn foo() {
    ChatStatus {};
}
```
becomes this
```rust
#[tracing::instrument]
pub fn foo() {
    ChatStatus { new_count: todo!(), last_msg: todo!() }};
                                                     // ^ extra closing brace 
}
```
Without `#[tracing::instrument]` expansion work as expected
